docs(cn): translate src/content/loaders/sass-loader.md (#788)
webpack-china/webpack.js.org: src/content/loaders/sass-loader.md

加载 Sass/SCSS 文件并将他们编译为 CSS。

## 快速开始

首先，你需要安装 `sass-loader`：

```console
npm install sass-loader sass webpack --save-dev
```

> `sass-loader` 需要预先安装 [Dart Sass](https://github.com/sass/dart-sass) 或 [Node Sass](https://github.com/sass/node-sass)（可以在这两个链接中找到更多的资料）。这可以控制所有依赖的版本， 并自由的选择使用的 Sass 实现。

Reviewer: ```suggestion
`sass-loader` 需要预先安装 [Dart Sass](https://github.com/sass/dart-sass) 或 [Node Sass](https://github.com/sass/node-sass)（可以在这两个链接中找到更多的资料）。
这可以控制所有依赖的版本，并自由的选择使用 Sass 的实现。
```

加载 Sass/SCSS 文件并将他们编译为 CSS。

## 快速开始

首先，你需要安装 `sass-loader`：

```console
npm install sass-loader sass webpack --save-dev
```

`sass-loader` 需要预先安装 [Dart Sass](https://github.com/sass/dart-sass) 或 [Node Sass](https://github.com/sass/node-sass)（可以在这两个链接中找到更多的资料）。这可以控制所有依赖的版本， 并自由的选择使用的 Sass 实现。

> 这样可以允许您控制所有依赖项的版本，并选择要使用的Sass实现。

Reviewer: ```suggestion
这样可以控制所有依赖项的版本，并选择要使用的 Sass 实现。
```

> > ℹ️ We recommend using [Dart Sass](https://github.com/sass/dart-sass).

Reviewer: 这里是更新的内容，需要翻译一下

};
```

最后通过你喜欢的方式运行 `webpack`。

> ### 解析 `import` at-rules

Reviewer: ```suggestion
### 解析 `import` 的规则
```

Webpack 提供一种 [解析文件的高级机制](/concepts/module-resolution/)。

`sass-loader` 使用 Sass 提供的 custom importer 特性，将所有 query 传递给 Webpack 解析引擎。只要在包名前加上 `~` ，告诉 Webpack 这不是一个相对路径，这样就可以从 `node_modules` 中 import 自己的 Sass 模块了：

```scss
@import '~bootstrap';
```

重要的是，只在前面加上 `~`，因为`~/` 将会解析到用户的主目录（home directory）。
因为 CSS 和 Sass 文件没有用于导入相关文件的特殊语法，所以 Webpack 需要区分 `bootstrap` 和 `~bootstrap`。
 `@import "style.scss"` 和 `@import "./style.scss";` 两种写法是相同的。

> ###  `url(...)` 的问题

Reviewer: ```suggestion
### `url(...)` 的问题
```

> 由于 Saass 的实现没有提供 [url rewriting](https://github.com/sass/libsass/issues/532) 的功能，所以相关的资源都必须是相对于输出文件（ouput）而言的。

Reviewer: ```suggestion
由于 Saass 的实现没有提供 [url 重写](https://github.com/sass/libsass/issues/532)的功能，所以相关的资源都必须是相对于输出文件（ouput）而言的。
```

- 如果生成的 CSS 传递给了 `css-loader`，则所有的 url 规则都必须是相对于入口文件的（例如：`main.scss`）。
- 如果仅仅生成了 CSS 文件，没有将其传递给 `css-loader`，那么所有的 url 都是相对于网站的根目录的。

第一种情况可能会带来一些困扰。通常情况下我们希望相对路径引用的解析是相对于声明它的 `.sass`/`.scss` 文件（如同在 `.css` 文件中一样）。

幸运的是，有两种方法可以解决这个问题：

- 将 [resolve-url-loader](https://github.com/bholloway/resolve-url-loader) 设置于 loader 链中的 `sass-loader` 之前，就可以重写 url。

> - Library 作者一般都会提供变量，用来设置资源路径。 比如 [bootstrap-sass](https://github.com/twbs/bootstrap-sass) 可以通过 `$icon-font-path` 进行设置。

Reviewer: ```suggestion
- Library 作者一般都会提供变量，用来设置资源路径。比如 [bootstrap-sass](https://github.com/twbs/bootstrap-sass) 可以通过 `$icon-font-path` 进行设置。
```

## 配置选项

|                   名称                    |         类型         |       默认值       | 描述                                           |
| :---------------------------------------: | :------------------: | :----------------: | :--------------------------------------------- |
|  **[`implementation`](#implementation)**  |      `{Object}`      |       `sass`       | 设置使用的 Sass 的实现。                       |
|     **[`sassOptions`](#sassoptions)**     | `{Object\|Function}` | Sass 实现的默认值  | Sass 自身选项。                                |
|       **[`sourceMap`](#sourcemap)**       |     `{Boolean}`      | `compiler.devtool` | 启用 / 禁用 source maps 的生成。               |

> |     **[`prependData`](#sassoptions)**     | `{String\|Function}` |    `undefined`     | 在实际的输入文件之前添加 `Sass` /`SCSS` 代码。 |

Reviewer: ```suggestion
|     **[`prependData`](#sassoptions)**     | `{String\|Function}` |    `undefined`     | 在实际的输入文件之前添加 `Sass`/`SCSS` 代码。 |
```

| **[`webpackImporter`](#webpackimporter)** |     `{Boolean}`      |       `true`       | 启用 / 禁用默认的 Webpack importer。           |

### `implementation`

类型： `Object`
默认值： `sass`

特殊的 `implementation` 选项确定要使用的 Sass 实现。 

默认情况下，loader 将会根据你的依赖解析需要使用的实现。
只需将必需的实现添加到 `package.json`（`sass` 或 `node-sass` 包）中并安装依赖项即可。

> 例如此时 `sass-loader` 将会使用 `sass` （`dart-sass`）实现：

Reviewer: ```suggestion
示例，此时 `sass-loader` 将会使用 `sass` （`dart-sass`）实现：
```

}
```

> 例如此时 `sass-loader` 将会使用 `node-sass` 实现：

Reviewer: ```suggestion
示例，此时 `sass-loader` 将会使用 `node-sass` 实现：
```

}
```

> 需要注意同时安装 `node-sass` 和 `sass` 的情况！ 默认情况下，`sass-loader` 会选择 `sass`。

Reviewer: ```suggestion
需注意同时安装 `node-sass` 和 `sass` 的情况！默认情况下，`sass-loader` 会选择 `sass`。
```

}
```

需注意同时安装 `node-sass` 和 `sass` 的情况！默认情况下，`sass-loader` 会选择 `sass`。

> 为了避免这种情况，您可以使用 `implementation` 选项。

Reviewer: ```suggestion
为了避免这种情况，你可以使用 `implementation` 选项。
```

};
```

需要注意的是，当使用 `sass`（`Dart Sass`）时，由于异步回调的开销，通常情况下**同步编译的速度是异步编译速度的两倍**。
为了避免这种开销，你可以使用 [fibers](https://www.npmjs.com/package/fibers) 包从同步代码中调用异步导入程序。

> 如果可能，我们会自动注入 [`fibers`](https://github.com/laverdet/node-fibers) 软件包（设置 `sassOptions.fiber`）（当然需要您安装 [`fibers`](https://github.com/laverdet/node-fibers) 包）。

Reviewer: ```suggestion
如果可能，我们会自动注入 [`fibers`](https://github.com/laverdet/node-fibers) 软件包（设置 `sassOptions.fiber`）（当然需要你安装 [`fibers`](https://github.com/laverdet/node-fibers) 包）。
```

### `sassOptions`

类型：`Object|Function`
默认值：Sass 实现的默认值

[Dart Sass](http://sass-lang.com/dart-sass) 或者 [Node Sass](https://github.com/sass/node-sass) 实现的选项。

> > ℹ️ `indentedSyntax` 选项对于 `sass` 拓展为 `true`。

Reviewer: ```suggestion
> ℹ️ `indentedSyntax` 选项值为 `true`，是对 `sass` 的扩展。
```

> > ℹ️ 像 `data` 和 `file` 这样的选项是不可用的并且他们将会被忽略。

Reviewer: ```suggestion
> ℹ️ 像 `data` 和 `file` 这样的选项是不可用的，且会被忽略。
```

### `sassOptions`

类型：`Object|Function`
默认值：Sass 实现的默认值

[Dart Sass](http://sass-lang.com/dart-sass) 或者 [Node Sass](https://github.com/sass/node-sass) 实现的选项。

> ℹ️ `indentedSyntax` 选项值为 `true`，是对 `sass` 的扩展。

> ℹ️ 像 `data` 和 `file` 这样的选项是不可用的，且会被忽略。

> ℹ 我们推荐不要设置 `outFile`，`sourceMapContents`，`sourceMapEmbed`，`sourceMapRoot` 这些选项，因为当 `sourceMap` 是 `true` 时，`sass-loader` 会自动设置这些选项。

> > ℹ️ Access to the [loader context](/api/loaders/#the-loader-context) inside the custom importer can be done using the `this.webpackLoaderContext` property.

Reviewer: 这里是更新的内容，翻译一下

> > ℹ️ 可以使用 `this.webpackLoaderContext` 属性访问 custom  importer 中的 [loader context](/api/loaders/#the-loader-context)

Reviewer: ```suggestion
> ℹ️ 可以使用 `this.webpackLoaderContext` 属性访问自定义 importer 中的 [loader 上下文](/api/loaders/#the-loader-context)。
```

### `sourceMap`

类型：`Boolean`
默认值：取决于 `compiler.devtool` 的值

> 开启 / 关闭生成 source maps。

Reviewer: ```suggestion
开启/关闭生成 source map。
```

### `prependData`

类型：`String|Function`
默认值：`undefined`

> 在实际的文件之前要添加的 `Sass` / `SCSS` 代码。

Reviewer: ```suggestion
在实际的文件之前要添加的 `Sass`/`SCSS` 代码。
```

在这种情况下，`sass-loader` 将不会覆盖 `data` 选项，而只是将它拼接在入口文件内容之前。

当某些 Sass 变量取决于环境时，这非常有用：

> > ℹ 由于代码代码，将破坏入口文件的 source mappings。 通常一个简单的解决方案，多个 Sass 入口文件。

Reviewer: ```suggestion
> ℹ 由于代码代码，将破坏入口文件的 source mappings。通常一个简单的解决方案，多个 Sass 入口文件。
```

### `webpackImporter`

类型：`Boolean`
默认值：`true`

开启 / 关闭默认的 Webpack importer。

> 在某些情况下，可以提高性能。 但是请谨慎使用，因为 aliases 和以 `〜` 开头的 `@import` 规则将不起作用。

Reviewer: ```suggestion
在某些情况下，可以提高性能。但是请谨慎使用，因为 aliases 和以 `〜` 开头的 `@import` 规则将不起作用。
```

};
```

## 示例

### 提取样式表

> 对于生产版本，我们建议从 bundle 中提取 CSS，以便之后可以使 CSS / JS 资源并行加载。

Reviewer: ```suggestion
对于生产版本，我们建议从 bundle 中提取 CSS，以便之后可以使 CSS/JS 资源并行加载。
```

> 从 bundle 中提取样式表，有2种可用的方法：

Reviewer: ```suggestion
从 bundle 中提取样式表，有 2 种实用的方式：
```

### Source maps

> 开启 / 关闭 source maps 的生成。

Reviewer: ```suggestion
开启/关闭 source map 的生成。
```

为了开启 CSS source maps，需要将 `sourceMap` 选项作为参数，传递给 `sass-loader` 和 `css-loader`。

**webpack.config.js**

```javascript
module.exports = {

>   devtool: 'source-map', // 任何类似于“source-map”的选项都是可以的

Reviewer: ```suggestion
  devtool: 'source-map', // 任何类似于 "source-map" 的选项都是支持的
```

};
```

> 如果你要在 Chrome 中编辑原始的 Sass 文件， 建议阅读 [这篇不错的博客](https://medium.com/@toolmantim/getting-started-with-css-sourcemaps-and-in-browser-sass-editing-b4daab987fb0)。具体示例参考 [test/sourceMap](https://github.com/webpack-contrib/sass-loader/tree/master/test)。

Reviewer: ```suggestion
如果你要在 Chrome 中编辑原始的 Sass 文件，建议阅读 [这篇不错的博客](https://medium.com/@toolmantim/getting-started-with-css-sourcemaps-and-in-browser-sass-editing-b4daab987fb0)。具体示例参考 [test/sourceMap](https://github.com/webpack-contrib/sass-loader/tree/master/test)。
```

> ## Contributing

Reviewer: ```suggestion
## 贡献
```
